clay/random.clj

```diff
@@ -1,11 +1,2274 @@
+      (interpose ",")))
+
+^:kindly/hide-code
+(def kde-kernels (:kde k/kernel-list))
+
+;; KDE kernels distribution plots
+
+(kind/table
+ (->> kde-kernels
+      (mapv (fn [k]
+              [(kind/hiccup
+                [:dl
+                 [:dt (kind/code (str k))]
+                 [:dd (kind/table [(mapv gg/->image (gg/dgraph (r/distribution :continuous-distribution {:data [-2 -2 -2 -1 0 1 2 -1 0 1 2 0 1 2 1 2 2]
+                                                                                                         :kde k :steps 100}) {:pdf [-3 3]}))])]])]))))
+
+
+;; #### Empirical
+
+;; Empirical distribution calculates PDF, CDF and iCDF from a histogram.
+
+;; * Name: `:empirical`
+;; * Default parameters
+;;    * `:data`, samples, sequence of numbers
+;;    * `:bin-count`, number of bins for histogram: 10% of the size of the data
+;; * [wiki](https://en.wikipedia.org/wiki/Empirical_distribution_function), [source](https://commons.apache.org/proper/commons-math/javadocs/api-3.6.1/org/apache/commons/math3/random/EmpiricalDistribution.html)
+
+(def empirical-distr (r/distribution :empirical {:data random-data}))
+
+(utls/dgraph-table
+ [["default", "bin-count=10"]
+  [(mapv gg/->image (gg/dgraph empirical-distr {:pdf [-5 5]}))
+   (mapv gg/->image (gg/dgraph (r/distribution :empirical {:data random-data :bin-count 10}) {:pdf [-5 5]}))]])
+
+(utls/examples-note
+ (r/sample empirical-distr)
+ (r/sample empirical-distr)
+ (r/pdf empirical-distr 0.0)
+ (r/cdf empirical-distr 0.0)
+ (map (partial r/icdf empirical-distr) [0.0001 0.5 0.9999])
+ (r/lower-bound empirical-distr)
+ (r/upper-bound empirical-distr)
+ (r/mean empirical-distr)
+ (r/variance empirical-distr))
+
+;; #### Discrete
+
+;; * Default parameters
+;;    * `:data`, sequence of numbers (integers/longs or doubles)
+;;    * `:probabilities`, optional, probabilities or weights
+;; * [wiki](https://en.wikipedia.org/wiki/Probability_distribution#Discrete_probability_distribution), [source1](https://commons.apache.org/proper/commons-math/javadocs/api-3.6.1/org/apache/commons/math3/distribution/EnumeratedRealDistribution.html), [source2](https://commons.apache.org/proper/commons-math/javadocs/api-3.6.1/org/apache/commons/math3/distribution/EnumeratedIntegerDistribution.html)
+
+;; Please note: data can contain duplicates.
+
+;; There are four discrete distributions:
+
+;; * `:enumerated-int` for integers, backed by Apache Commons Math
+;; * `:enumerated-real` for doubles, backed by Apache Commons Math
+;; * `:integer-discrete-distribution` - for longs, custom implementation
+;; * `:real-discrete-distribution` - for doubles, custom implementation
+
+;; Please note:
+
+;; * Apache Commons Math implementation have some issues with iCDF.
+;; * `:integer-discrete-distribution` is backed by `clojure.data.int-map`
+
+;; ##### Doubles
+
+(def data-doubles (repeatedly 100 #(m/sq (m/approx (r/drand 2.0) 1))))
+
+(let [enumerated-real (r/distribution :enumerated-real {:data data-doubles})
```

Comment:
Why is it packed into this hiccup structure? Is it necessary? Can we implement this part as other distribution examples, ie. just in tables?

Reply:
For all distribution charts, including those shown above, I organize the param-chart pairs in a single-column table structure. Each pair occupies its row within this table. To enhance semantic meaning, I utilize the HTML <dl> (description list) tag, implemented through a hiccup structure. This approach clearly distinguishes between the parameters and their corresponding charts.

This table-hiccup structure is consistently applied across all distribution examples rendered using the `fastmath.dev.clay/dgraph-table` macro.

Reply:
Ah, ok. So my question is: can we use this macro here as well?

Reply:
Yes, I've changed the code to use `dgraph-table` macro.